code_method/code_method/bootstrap_test_compLasso_rf.R

```diff
@@ -0,0 +1,127 @@
+##########################################################
+### estimate correlation between stability index #########
+##########################################################
+
+source('cv_method.R') 
+source('getStability.R')
+source('cv_sim_apply.R')
+
+## set up parallel computing
+library(foreach)
+library(doParallel)
+numCores <- detectCores() - 2 # 6 cores
+registerDoParallel(numCores)  # use multicore, set to the number of our cores
+
+boot_stab_sim = function(num_boot=100, sim_file, method, seednum=31, ratio.training=0.8, fold.cv=10, 
+					  family='gaussian', lambda.grid=exp(seq(-4, -2, 0.2)), alpha.grid=seq(0.1, 0.9, 0.1), 
+	                  mtry.grid=seq(5, 25, 5), num_trees = 500, pval_thr = 0.05){
+	
+	# load simulated data
+	load(sim_file, dat <- new.env())
+
+	idx.start = 1; idx.stop = 100 # 100 repetitions for each simulated scenario
+	rou = dat$sim_array[[1]]$rou # rou, n, p are same across all repetitions
+	n = dat$sim_array[[1]]$n
+	p = dat$sim_array[[1]]$p
+
+	## get a vector of stability index from bootstrapped data
+	stab_index = rep(0, num_boot)
+	b = 0
+	for (i in idx.start:idx.stop){
+		b = b + 1
+		print(paste('index', i, sep=':'))
+	    sub = dat$sim_array[[i]]
+
+	    # bootsrap with parallelization
+		selections = foreach (i=1:num_boot) %dopar% {
+			N = length(sub$Y) # number of samples
+			boot_ids = sample(N, size=N, replace=TRUE)
+			boot_Z = sub$Z[boot_ids, ]
+			boot_Y = sub$Y[boot_ids, ]
+
+		    ## select features from lasso/elnet
+		    if (method == 'compLasso'){
+		        result.lin = cons_lasso_cv(y=boot_Y, datx=boot_Z, seednum=i, ratio.training=ratio.training)
+		        select.lin = result.lin$coef.chosen  # since 1 represents intercept
+
+		    } else if (method == 'RF'){
+		        result.rf = randomForest_cv(y=boot_Y, datx=boot_Z, seednum=i, fold.cv=fold.cv, 
+	    							        num_trees=num_trees, mtry.grid = mtry.grid, pval_thr=pval_thr)
+	            select.rf = result.rf$coef.chosen 
+		    }
+		}
+
+		# calculate stability index from bootstrapped data
+		stability_table = matrix(rep(0, num_boot * p), ncol=p)
+		for (j in 1:num_boot){
+		stability_table[j, selections[[j]]] = 1 
+		}
+
+		stab_index[b] = round(getStability(stability_table)$stability, 2)
```

Comment:
You get NA values because the entire X matrix are filled with zeros (indicating no features were chosen in any bootstrapped samples), when stability is not defined. Actually if you change the seed, try set.seed(1), 2, 3, then you will find that it is only NA for seed 2, when X is all zeros, while you would get a small number close to 0 for seed 1 or 3, when there in only one feature being selected in one bootstrapped data. In other words, Stability allows for high sparsity, as long as there is one 1 that's sufficient, but not for all zeros (which was your case).